### PR TITLE
[X86][MC] Add R_X86_64_CODE_4_GOTTPOFF

### PR DESCRIPTION
**Pull request by fzou1**
For

  mov name@GOTTPOFF(%rip), %reg
  add name@GOTTPOFF(%rip), %reg

add

  `R_X86_64_CODE_4_GOTTPOFF` = 44

if the instruction starts at 4 bytes before the relocation offset. It's similar to R_X86_64_GOTTPOFF.

Linker can treat `R_X86_64_CODE_4_GOTTPOFF` as `R_X86_64_GOTTPOFF` or convert the instructions above to

  mov $name@tpoff, %reg
  add $name@tpoff, %reg

if the first byte of the instruction at the relocation `offset - 4` is `0xd5` (namely, encoded w/REX2 prefix) when possible.

Binutils patch: https://github.com/bminor/binutils-gdb/commit/a533c8df598b5ef99c54a13e2b137c98b34b043c
Binutils mailthread: https://sourceware.org/pipermail/binutils/2023-December/131463.html
ABI discussion: https://groups.google.com/g/x86-64-abi/c/ACwD-UQXVDs/m/vrgTenKyFwAJ
Blog: https://kanrobert.github.io/rfc/All-about-APX-relocation

### CODE REVIEW

File: llvm/include/llvm/BinaryFormat/ELFRelocs/x86_64.def [hunk 1]
```diff
@@ -44,3 +44,4 @@ ELF_RELOC(R_X86_64_IRELATIVE,   37)
 ELF_RELOC(R_X86_64_GOTPCRELX,   41)
 ELF_RELOC(R_X86_64_REX_GOTPCRELX,    42)
 ELF_RELOC(R_X86_64_REX2_GOTPCRELX,    43)
+ELF_RELOC(R_X86_64_CODE_4_GOTTPOFF,    44)
```
Comment (MaskRay) [inline llvm/include/llvm/BinaryFormat/ELFRelocs/x86_64.def:47]: Missing trailing \n

File: llvm/test/MC/X86/tls.s [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+// CHECK-NEXT:   addq (%rip), %r16
+// CHECK-NEXT:   R_X86_64_CODE_4_GOTTPOFF    tls0-0x4
+// CHECK-NEXT:   addq (%rip), %r28
+// CHECK-NEXT:   R_X86_64_CODE_4_GOTTPOFF    tls0-0x4
+// CHECK-NEXT:   addq (%rip), %r16
+// CHECK-NEXT:   R_X86_64_CODE_4_GOTTPOFF    tls1-0x4
+
+.type tls0,@object
+.section .tbss,"awT",@nobits
+.globl tls0
```
Comment (MaskRay) [inline llvm/test/MC/X86/tls.s:20]: drop `.align` and `.size`, unneeded by this test

File: llvm/test/MC/X86/tls.s [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+.section .text
+.globl _start
+_start:
+    # EGPR
+    movq tls0@GOTTPOFF(%rip), %r16
+    movq tls0@GOTTPOFF(%rip), %r20
+    movq tls1@GOTTPOFF(%rip), %r16
+    addq tls0@GOTTPOFF(%rip), %r16
+    addq tls0@GOTTPOFF(%rip), %r28
+    addq tls1@GOTTPOFF(%rip), %r16
```
Comment (MaskRay) [inline llvm/test/MC/X86/tls.s:40]: missing trailing \n

Comment (fzou1) [reply llvm/test/MC/X86/tls.s]: Done

File: llvm/test/MC/X86/tls.s (hunk 2 above)
Comment (fzou1) [reply llvm/test/MC/X86/tls.s:20]: Done.

File: llvm/include/llvm/BinaryFormat/ELFRelocs/x86_64.def (hunk 1 above)
Comment (fzou1) [reply llvm/include/llvm/BinaryFormat/ELFRelocs/x86_64.def]: Done. Thanks.

File: llvm/test/MC/X86/tls.s [hunk 4]
```diff
@@ -0,0 +1,37 @@
+// RUN: llvm-mc -filetype=obj -triple=x86_64 %s -o %t
+// RUN: llvm-objdump -dr --no-print-imm-hex %t | FileCheck %s
+
+// CHECK:      <_start>:
```
Comment (KanRobert) [inline llvm/test/MC/X86/tls.s:4]: Why not use the existing test file llvm/test/MC/ELF/relocation.s?

Comment (fzou1) [reply llvm/test/MC/X86/tls.s:4]: There were no checks for both instructions and relocation types emitted by llvm-mc, so I added new test file for this R_X86_64_CODE_4_GOTTPOFF type and other relocation types to be added.

Comment (KanRobert) [reply llvm/test/MC/X86/tls.s:4]: Why do we need to check the instructions? There should be existing encoding/decoding tests there.

IMHO, checking the similar relocation types in a single file can help use not miss some checks?

Comment (fzou1) [reply llvm/test/MC/X86/tls.s:4]: It's not to check encoding/decoding. It showed the map of instruction and relocation information (type, symbol and offset), and that's the output what users got from llvm-mc/llvm. Some tests in llvm/test/MC folder did same thing.

I'm okay with adding the check of relocation information in relocation.s and not adding this test file, if you insist.

Comment (fzou1) [reply llvm/test/MC/X86/tls.s]: I've removed the test file newly added and add the check in relocation.s.

**Review by KanRobert — APPROVED** [review]
LGTM

**Comment by fzou1** [comment]
@MaskRay, any comments?

If no objection, @KanRobert /@MaskRay, please help merge this PR. I don't have commit access.

Thanks.

**Review by MaskRay — APPROVED** [review]

**Pull request merged**